internal/controller/nonadminrestore_controller.go

```diff
@@ -0,0 +1,339 @@
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	nacv1alpha1 "github.com/migtools/oadp-non-admin/api/v1alpha1"
+	"github.com/migtools/oadp-non-admin/internal/common/constant"
+	"github.com/migtools/oadp-non-admin/internal/common/function"
+	"github.com/migtools/oadp-non-admin/internal/handler"
+	"github.com/migtools/oadp-non-admin/internal/predicate"
+)
+
+// NonAdminRestoreReconciler reconciles a NonAdminRestore object
+type NonAdminRestoreReconciler struct {
+	client.Client
+	Scheme        *runtime.Scheme
+	OADPNamespace string
+}
+
+type nonAdminRestoreReconcileStepFunction func(ctx context.Context, logger logr.Logger, nar *nacv1alpha1.NonAdminRestore) (bool, error)
+
+//+kubebuilder:rbac:groups=oadp.openshift.io,resources=nonadminrestores,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=oadp.openshift.io,resources=nonadminrestores/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=oadp.openshift.io,resources=nonadminrestores/finalizers,verbs=update
+
+// +kubebuilder:rbac:groups=velero.io,resources=restores,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state,
+// defined in NonAdminRestore object Spec.
+func (r *NonAdminRestoreReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	logger.V(1).Info("NonAdminRestore Reconcile start")
+
+	nar := &nacv1alpha1.NonAdminRestore{}
+	err := r.Get(ctx, req.NamespacedName, nar)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.V(1).Info(err.Error())
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "Unable to fetch NonAdminRestore")
+		return ctrl.Result{}, err
+	}
+
+	if !nar.DeletionTimestamp.IsZero() {
+		updatedPhase := updateNonAdminPhase(&nar.Status.Phase, nacv1alpha1.NonAdminPhaseDeleting)
+		updatedCondition := meta.SetStatusCondition(&nar.Status.Conditions,
+			metav1.Condition{
+				Type:    string(constant.NonAdminConditionDeleting),
+				Status:  metav1.ConditionTrue,
+				Reason:  "DeletionPending",
+				Message: "restore accepted for deletion",
+			},
+		)
+		if updatedPhase || updatedCondition {
+			if err := r.Status().Update(ctx, nar); err != nil {
+				logger.Error(err, statusUpdateError)
+				return ctrl.Result{}, err
+			}
+			logger.V(1).Info("NonAdminRestore status marked for deletion")
+		} else {
+			logger.V(1).Info("NonAdminRestore status unchanged during deletion")
+		}
+
+		veleroRestore, err := function.GetVeleroRestoreByLabel(ctx, r.Client, r.OADPNamespace, nar.Status.UUID)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				logger.Error(err, "Unable to fetch Velero Restore")
+				return ctrl.Result{}, err
+			}
+		} else {
+			// 	TODO when/why Velero Restore has finalizer?
```

Comment:
I think velero docs are wrong

I tested running normal delete (`oc delete`) on velero restore objcet, and things in storage where deleted

Reply:
I've also tested this, so yeah seems like there is something fishy with the docs, however I've played with the restore a bit more and the finalizer is only removed when actually logs are removed:



Success on restore:
```shell
$ aws s3 ls  MY-BUCKET/default-nac-bucket/restores/mongo-persistent-restore/
2024-11-19 12:52:12         27 mongo-persistent-restore-volumeinfo.json.gz
2024-11-19 12:52:12         27 restore-mongo-persistent-restore-itemoperations.json.gz
2024-11-19 12:52:12       7535 restore-mongo-persistent-restore-logs.gz
2024-11-19 12:52:12        615 restore-mongo-persistent-restore-resource-list.json.gz
2024-11-19 12:52:12        331 restore-mongo-persistent-restore-results.gz

$ oc delete restore  mongo-persistent-restore -n openshift-adp
restore.velero.io "mongo-persistent-restore" deleted

$ aws s3 ls MY-BUCKET/default-nac-bucket/restores/mongo-persistent-restore/
<empty>
```

Failure to delete s3 storage on restore (read only permissions):
**IAM_NUMBER**, **MY-USERNAME** and **MY-BUCKET** needs to be properly set:

```shell
$ cat /tmp/disable-delete-bucket.json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Deny",
      "Action": "s3:DeleteObject",
      "Resource": "arn:aws:s3:::MY-BUCKET/*",
      "Condition": {
        "DateGreaterThan": {
          "aws:CurrentTime": "2024-11-19T00:00:00Z"
        },
        "DateLessThan": {
          "aws:CurrentTime": "2024-11-19T23:59:59Z"
        }
      },
      "Principal": {
        "AWS": "arn:aws:iam::IAM_NUMBER:user/MY-USERNAME"
      }
    }
  ]
}


$ aws s3api put-bucket-policy --bucket MY-BUCKET --policy file:///tmp/disable-delete-bucket.json
```

After the above steps the mongo-persistent-restore was still in the cluster, because finalizer was preventing it from being fully removed. After manually removing finalizer I was able to see that object being gone from the cluster, however as expected s3 restore storage was intact. So I think we should somehow see if the restore object is really gone and only then remove finalizer from the Non Admin restore object or we are risking having orphant restore object without user knowing it.

```shell
$ oc delete restore  mongo-persistent-restore -n openshift-adp
restore.velero.io "mongo-persistent-restore" deleted

$ oc describe restore mongo-persistent-restore -n openshift-adp
Name:         mongo-persistent-restore
Namespace:    openshift-adp
Labels:       <none>
Annotations:  <none>
API Version:  velero.io/v1
Kind:         Restore
Metadata:
  Creation Timestamp:             2024-11-19T11:55:37Z
  Deletion Grace Period Seconds:  0
  Deletion Timestamp:             2024-11-19T11:59:57Z
  Finalizers:
    restores.velero.io/external-resources-finalizer
  Generation:  7
  Managed Fields:
    API Version:  velero.io/v1
    Fields Type:  FieldsV1
    fieldsV1:
      f:spec:
        .:
        f:backupName:
    Manager:      Mozilla
    Operation:    Update
    Time:         2024-11-19T11:55:37Z
    API Version:  velero.io/v1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:finalizers:
          .:
          v:"restores.velero.io/external-resources-finalizer":
      f:spec:
        f:excludedResources:
        f:itemOperationTimeout:
      f:status:
        .:
        f:completionTimestamp:
        f:hookStatus:
        f:phase:
        f:progress:
          .:
          f:itemsRestored:
          f:totalItems:
        f:startTimestamp:
        f:warnings:
    Manager:         velero-server
    Operation:       Update
    Time:            2024-11-19T11:55:40Z
  Resource Version:  5887493
  UID:               090e7342-c162-40ed-b4cf-138819f99805
Spec:
  Backup Name:  mongo-persistent-backup
  Excluded Resources:
    nodes
    events
    events.events.k8s.io
    backups.velero.io
    restores.velero.io
    resticrepositories.velero.io
    csinodes.storage.k8s.io
    volumeattachments.storage.k8s.io
    backuprepositories.velero.io
  Item Operation Timeout:  4h0m0s
Status:
  Completion Timestamp:  2024-11-19T11:55:40Z
  Hook Status:
  Phase:  Completed
  Progress:
    Items Restored:  34
    Total Items:     34
  Start Timestamp:   2024-11-19T11:55:37Z
  Warnings:          18
Events:              <none>
```